File: src/awaitable_event/awaitable_event.erl
-module(awaitable_event).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../../docs/awaitable_event.md"}.

-export([
    run/1
]).

-spec run(WorkerStartDelay :: non_neg_integer()) -> ok.
run(WorkerStartDelay) ->
    spawn(fun() -> start_activity_worker(WorkerStartDelay) end),
    temporal_sdk:start_workflow(cluster_1, "default", awaitable_event_workflow, [wait]),
    ok.

start_activity_worker(WorkerStartDelay) ->
    case temporal_sdk_worker:is_started(cluster_1, activity, awaitable_event_activity) of
        false ->
            timer:sleep(WorkerStartDelay),
            io:fwrite("Start activity worker.~n", []),
            {ok, #{}} = temporal_sdk_worker:start(cluster_1, activity, [
                {task_queue, "awaitable_event"}, {worker_id, awaitable_event_activity}
            ]);
        true ->
            io:fwrite("Activity worker already started.~n", [])
    end.

File: lib/awaitable_event/awaitable_event.ex
defmodule AwaitableEvent do
  @external_resource "docs/awaitable_event.md"
  @moduledoc TemporalSdk.Utils.exdoc!("docs/awaitable_event.md")

  @spec run(worker_start_delay :: non_neg_integer()) :: :ok
  def run(worker_start_delay) do
    spawn(fn -> start_activity_worker(worker_start_delay) end)
    TemporalSdk.start_workflow(:cluster_1, "default", AwaitableEvent.Workflow, [:wait])
    :ok
  end

  @doc false
  def start_activity_worker(worker_start_delay) do
    case TemporalSdk.Worker.is_started(:cluster_1, :activity, :awaitable_event_activity) do
      false ->
        Process.sleep(worker_start_delay)
        IO.puts("Start activity worker.")

        {:ok, %{}} =
          TemporalSdk.Worker.start(:cluster_1, :activity,
            task_queue: "awaitable_event",
            worker_id: :awaitable_event_activity
          )

      true ->
        IO.puts("Activity worker already started.")
    end
  end
end
